txn.go

```diff
@@ -184,12 +177,50 @@ func (o *oracle) newCommitTs(txn *Txn) uint64 {
 		ts = txn.commitTs
 	}
 
-	for _, w := range txn.writes {
-		o.commits[w] = ts // Update the commitTs.
+	if ts > o.lastCleanupTs {
+		o.committedTxns = append(o.committedTxns, committedTxn{
+			ts:     ts,
+			writes: txn.writes,
+		})
 	}
+
 	return ts
 }
 
+func (o *oracle) doneRead(txn *Txn) {
+	if !txn.doneRead {
+		txn.doneRead = true
+		o.readMark.Done(txn.readTs)
+	}
+}
+
+func (o *oracle) cleanupCommittedTransactions() { // Must be called under o.Lock
+	// Same logic as discardAtOrBelow but unlocked
+	var maxReadTs uint64
+	if o.isManaged {
+		maxReadTs = o.discardTs
+	} else {
+		maxReadTs = o.readMark.DoneUntil()
+	}
+
+	if maxReadTs <= o.lastCleanupTs {
```

Comment:
I'm curious how we could end up in this condition. `o.lastCleanupTs` should always be less than the `maxReadTs`.

Reply:
The `<` part is just defensive programming. I can replace that with an assert if you prefer.

The `==` part is an optimization: do not run clean up if the `maxReadTs` (which is the read timestamp of the oldest transaction that is still in flight) has not increased.

Reply:
>The < part is just defensive programming. I can replace that with an assert if you prefer.

I understand your point but I think we should complain here with a `y.Assert`. 

>The == part is an optimization: do not run clean up if the maxReadTs (which is the read timestamp of the oldest transaction that is still in flight) has not increased.

Oh, yes. That makes sense. Thanks.